### PR TITLE
Core data: useEntityBlockEditor: fix parsed blocks cache

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

In #58146, I didn't account for editing being made without the blocks key. This can happen when you change the content in the code editor, in which case the content needs to be parsed. Currently the cache will return the unedited blocks.

The solution is to change the cache key when edits are made. We can use the unedited record as a cache key when `getEntityRecordEdits` returns undefined or an empty object. I'm not sure why it sometimes returns an empty object. That sounds like a bug that needs to be fixed.

### CODE REVIEW

File: packages/core-data/src/entity-provider.js [hunk 1]
```diff
@@ -187,16 +187,28 @@ export function useEntityBlockEditor( kind, name, { id: _id } = {} ) {
...
+		// If not, cache by the original enity record.
+		const edits = getEntityRecordEdits( kind, name, id );
+		const isUnedited = ! edits || ! Object.keys( edits ).length;
+		const cackeKey = isUnedited ? getEntityRecord( kind, name, id ) : edits;
+		let _blocks = parsedBlocksCache.get( cackeKey );
 
 		if ( ! _blocks ) {
 			_blocks = parse( content );
-			parsedBlocksCache.set( entityRecord, _blocks );
+			parsedBlocksCache.set( cackeKey, _blocks );
```
Comment (youknowriad) [inline packages/core-data/src/entity-provider.js:199]: This might leave in memory two block lists unnecessarily. (one of entity record, one for the latest edits object). So there's a tiny memory impact, probably ok though.

File: test/e2e/specs/editor/various/editor-modes.spec.js [hunk 2]
```diff
@@ -155,4 +155,35 @@ test.describe( 'Editing modes (visual/HTML)', () => {
...
+	} ) => {
+		// Open the code editor.
+		await pageUtils.pressKeys( 'secondary+M' );
+
+		const textbox = page.getByRole( 'textbox', {
+			name: 'Type text or HTML',
+		} );
+
+		// Change content by typing.
+		await textbox.fill( '' );
```
Comment (youknowriad) [inline test/e2e/specs/editor/various/editor-modes.spec.js:172]: Why is this necessary?

**Review by youknowriad — APPROVED** [review]

File: packages/core-data/src/entity-provider.js (hunk 1 above)
Comment (ellatrix) [reply packages/core-data/src/entity-provider.js:199]: I think that's fine. Parts of this block list will be reused for edited block lists I think?

File: test/e2e/specs/editor/various/editor-modes.spec.js (hunk 2 above)
Comment (ellatrix) [reply test/e2e/specs/editor/various/editor-modes.spec.js:172]: There's a beforeEach that fills in content

**Review by SantosGuillamot — APPROVED** [review]
I've tested it and it seems to solve the issue 🙂

**Comment by youknowriad** [comment]
Can we rebase and merge this PR?

File: packages/core-data/src/entity-provider.js (hunk 1 above)
Comment (adamziel) [inline packages/core-data/src/entity-provider.js:199]:
```suggestion
			parsedBlocksCache.set( cacheKey, _blocks );
```

File: test/e2e/specs/editor/various/editor-modes.spec.js [hunk 3]
```diff
@@ -155,4 +155,35 @@ test.describe( 'Editing modes (visual/HTML)', () => {
...
+		await textbox.focus();
+
+		await page.keyboard.type( `<!-- wp:paragraph -->
+<p>abc</p>
+<!-- /wp:paragraph -->` );
```
Comment (Mamaduka) [inline test/e2e/specs/editor/various/editor-modes.spec.js:177]: @ellatrix, is typing action here intentional or can we just use `fill`?

I'm asking because typing produces block validation warnings and creates a lot of noise in e2e test reports, which can hide the real issues.

Comment (ellatrix) [reply test/e2e/specs/editor/various/editor-modes.spec.js:177]: It should be fine if the test still passes

Comment (Mamaduka) [reply test/e2e/specs/editor/various/editor-modes.spec.js:177]: Created a follow-up https://github.com/WordPress/gutenberg/pull/62957.
